src/components/chip/chip.e2e.ts

```diff
@@ -19,10 +19,22 @@ describe('limel-chip', () => {
         await page.setContent(
             '<limel-chip text="Test chip" link="{ href: \'https://example.com\' }"></limel-chip>',
         );
+
+        const element = await page.find('limel-chip');
+        element.setProperty('link', { href: 'https://example.com' });
+
         await page.waitForChanges();
 
-        const element = await page.find('limel-chip >>> a.chip');
-        expect(element.getAttribute('href')).toEqual('https://example.com');
+        expect(element)
```

Comment:
The problem is that the shadowRoot is an empty dict. It seems that in other places we've opted to just use the HTML.

You might wonder why we need to invoke `setProperty`. I don't know. It just breaks if I don't do it here, even if we wait for changes. Manually everything works.

But then it also breaks if we don't put the link at the top even if we set th e property 🤷🏽‍♂️

Reply:
> The problem it seems is that the shadowRoot is an empty dict. It seems that in other places we've opted to just use the HTML where there was a shadow node.

Hmm… that's odd. In Jest (which .spec-files are run in) it makes sense, since Jest isn't running a real browser environment. But it's weird that there wouldn't be a real shadowRoot in e2e-tests, since those are run in a real browser by Puppeteer…

Although, the version of Puppeteer that's being used is some old version that still runs Chrome's old headless mode, which is apparently a completely separate browser implementation from the normal Chrome browser (which kind of defeats the whole purpose a bit, one would think 🤦‍♂️). That's really old, and is being replaced by a new headless mode that actually uses the regular Chrome browser. So, maybe that old headless mode just doesn't support shadow-DOM properly, and our tests are being so weird and hard to get to work because it's actually running some polyfill…? 🤔

> You might wonder why we need to invoke `setProperty`. I don't know. It just breaks if I don't do it here, even if we wait for changes. Manually everything works.

The reason is that the `Link` property is not a string. You can only pass strings as html attributes. Stencil magics it so that you seem to be able to pass anything when you do it in tsx, but here it's vanilla html, so it doesn't work.

The reason you have to use `.setProperty` rather than just do `element.link = { href: 'https://example.com/' }` is that the `element` isn't actually a DOM-node, it's a puppeteer object with references to the actual DOM-node.

> But then it also breaks if we don't put the link at the top even if we set the property 🤷🏽‍♂️

I don't understand what you mean by "put the link at the top"? 🤔

Reply:
> ...maybe that old headless mode just doesn't support shadow-DOM properly...

Maybe the old one doesn't, that seems reasonable 🤔. Interesting stuff. I'll dig around for a bit, now I'm curious how exactly they implement their e2e test framework

The "link at the top" is the link attribute inside the chip element defined at the top when we initialise the page when we arrange this test. 😅 I should've probably used a more precise descriptor

Reply:
> > ...maybe that old headless mode just doesn't support shadow-DOM properly...
> 
> Maybe the old one doesn't, that seems reasonable 🤔. Interesting stuff. I'll dig around for a bit, now I'm curious how exactly they implement their e2e test framework

I think that when you run the e2e-tests, there is some info logged about the new mode, and that you can opt-in to it by using some config. Perhaps we can simply opt-in to the new mode? I'm not sure how to though, because I think Stencil "hides" a bunch of the Puppeteer config from us, so I'm not sure if we can set that property that needs setting… 🤷‍♂️

> The "link at the top" is the link attribute inside the chip element defined at the top when we initialise the page when we arrange this test. 😅 I should've probably used a more precise descriptor

Ah! Now I get what you meant. No worries, I'm just pretty rubbish at figuring out what people mean sometimes! 🙈😅

That gave me in inkling actually. The chip looks for the `this.link` property when it renders, and renders a button if the link property is falsy. It ought to swap that for an anchor element when the link property becomes truthy, but if we've already tried to find that element by then, or the page doesn't have time to update before we do, then we're not going to find it.